File: java-kata/src/main/java/com/leetcode/DiameterOfBinaryTree.java
package com.leetcode;

import com.leetcode.data.TreeNode;

/**
 * <h3><a href="https://leetcode.com/problems/diameter-of-binary-tree/">Diameter of Binary Tree</a></h3>
 * <p>
 *     Given the <code>root</code> of a binary tree, return <em>the length of the <strong>diameter</strong> of the tree</em>.
 * </p>
 * <p>
 *     The <strong>diameter</strong> of a binary tree is the <strong>length</strong> of the longest path between
 *     any two nodes in a tree. This path may or may not pass through the <code>root</code>.
 * </p>
 * <p>
 *     The <strong>length</strong> of a path between two nodes is represented by the number of edges between them.
 * </p>
 * <p>
 *     <h6>Constraints:</h6>
 *     <ul>
 *         <li>The number of nodes in the tree is in the range <code>[1, 10^4]</code></li>
 *         <li><code>-100 <= Node.val <= 100</code></li>
 *     </ul>
 * </p>
 */
public class DiameterOfBinaryTree {

    public int diameterOfBinaryTree(TreeNode root) {
        return lookUp(root).max();
    }

    private Result lookUp(TreeNode root) {
        if (root == null) return new Result(0, 0);
        var leftResult = lookUp(root.left);
        var rightResult = lookUp(root.right);
        var subtreeMax = Math.max(rightResult.max, leftResult.max);
        var max = Math.max(subtreeMax, leftResult.chain + rightResult.chain);
        return new Result(max, Math.max(leftResult.chain, rightResult.chain) + 1);
    }

    private record Result(int max, int chain) {
    }
}

File: kotlin-kata/src/main/kotlin/com/leetcode/DiameterOfBinaryTree.kt
package com.leetcode

import com.leetcode.data.TreeNode
import kotlin.math.max

/**
 * ### [Diameter of Binary Tree](https://leetcode.com/problems/diameter-of-binary-tree/)
 *
 * Given the `root` of a binary tree, return *the length of the **diameter** of the tree*.
 *
 * The **diameter** of a binary tree is the **length** of the longest path between
 * any two nodes in a tree. This path may or may not pass through the `root`.
 *
 * The **length** of a path between two nodes is represented by the number of edges between them.
 *
 * ###### Constraints:
 *  - The number of nodes in the tree is in the range `[1, 10^4]`
 *  - `-100 <= Node.val <= 100`
 */
class DiameterOfBinaryTree {

    fun diameterOfBinaryTree(root: TreeNode?): Int {
        return lookUp(root).max
    }

    private fun lookUp(root: TreeNode?): Result {
        if (root == null) {
            return Result(0, 0)
        }
        val leftResult = lookUp(root.left)
        val rightResult = lookUp(root.right)
        val subtreeMax = max(leftResult.max, rightResult.max)
        val max = max(subtreeMax, leftResult.chain + rightResult.chain)
        return Result(max, max(leftResult.chain, rightResult.chain) + 1)
    }

    private data class Result(val max: Int, val chain: Int)
}
